### PR TITLE
work for #198, adding in a qr code reader

### PR DESCRIPTION
**Pull request by aliu139**
creating PR so I don't forget about this later. Probably needs some css and js cleanup + someone who actually knows jade

### CODE REVIEW

File: views/admin/checkin.jade [hunk 1]
```diff
@@ -7,9 +7,19 @@ block append vars
     - bodyAttr = {"data-ng-app": "brh.controllers"}
 
 block append admin_content
+    script(src='//cirocosta.github.io/qcode-decoder/build/qcode-decoder.min.js')
     div(data-ng-controller='checkin.ctrl')
         +contentHeading("participant checkin")
         #searchparameters.col-md-12
+            section#camera(ng-init='showVideo = true')
+                        label(style="display: block;text-align: center;") Scan QR Code
```
Comment (mahsu) [inline views/admin/checkin.jade]: move inline css properties to file

File: views/admin/checkin.jade [hunk 2]
```diff
@@ -34,6 +44,28 @@ block append admin_content
                             .checkbox.np-enabled
                                 label
                                     input.checkbox(type='checkbox', value='true', ng-change='checkinUser(user.pubid)', ng-model='checkin')
+    script.
```
Comment (mahsu) [inline views/admin/checkin.jade]: move inline js to file

File: views/admin/checkin.jade [hunk 3]
```diff
@@ -34,6 +44,28 @@ block append admin_content
...
+        }];
+        elems.forEach(function(e) {
+            e.activator.onclick = function(r) {
+                r && r.preventDefault(),
+                e.decoder.call(qr, e.target, function(e, r) {
+                    if (e)
+                        throw e;
+                    alert("Just decoded: " + r);
+                    location.reload();
+                }, !0)
```
Comment (mahsu) [inline views/admin/checkin.jade]: lol unminify the `!0` to true

File: views/admin/checkin.jade [hunk 4]
```diff
@@ -34,6 +44,28 @@ block append admin_content
...
+            throw alert("Your browser doesn't match the required specs."),
+            new Error("Canvas and getUserMedia are required");
+        var elems = [{
+            target: document.querySelector("#camera video"),
+            activator: document.querySelector("#camera button"),
+            decoder: qr.decodeFromCamera
+        }];
+        elems.forEach(function(e) {
+            e.activator.onclick = function(r) {
+                r && r.preventDefault(),
```
Comment (mahsu) [inline views/admin/checkin.jade]: I'm unfamiliar with the comma syntax for ` r && r.preventDefault(),` what is this supposed to do?

File: views/admin/checkin.jade [hunk 5]
```diff
@@ -34,6 +44,28 @@ block append admin_content
...
+                e.decoder.call(qr, e.target, function(e, r) {
+                    if (e)
+                        throw e;
```
Comment (mahsu) [inline views/admin/checkin.jade]: wrap ` throw e` in braces

File: views/admin/checkin.jade [hunk 6]
```diff
@@ -34,6 +44,28 @@ block append admin_content
                             .checkbox.np-enabled
                                 label
                                     input.checkbox(type='checkbox', value='true', ng-change='checkinUser(user.pubid)', ng-model='checkin')
+    script.
+        var qr = new QCodeDecoder;
+        var video = document.getElementById('camera');
+        if (!qr.isCanvasSupported() || !qr.hasGetUserMedia())
+            throw alert("Your browser doesn't match the required specs."),
```
Comment (mahsu) [inline views/admin/checkin.jade]: missing `{}` for lines 51-52?

File: views/admin/checkin.jade (hunk 1 above)
Comment (aliu139) [inline views/admin/checkin.jade]: Moved

File: views/admin/checkin.jade (hunk 2 above)
Comment (aliu139) [inline views/admin/checkin.jade]: Moved

File: views/admin/checkin.jade (hunk 6 above)
Comment (aliu139) [inline views/admin/checkin.jade]: Added

File: views/admin/checkin.jade (hunk 3 above)
Comment (aliu139) [inline views/admin/checkin.jade]: Done

**Comment by aliu139** [comment]
Addressed all comments in Matt's code review

**Comment by TrevorEdwards** [comment]
It's not really clear how this works- could you explain?

**Comment by aliu139** [comment]
Added comments to explain

**Comment by TrevorEdwards** [comment]
Err I actually meant how it should be used- as in if a volunteer is checking someone in they should be able to understand how the interface works from looking at it

**Comment by aliu139** [comment]
Oh lol. When you navigate to the checkin page, you just hit begin scanning. Then you point the camera at the qr code, and it should scan. The logic for actual checkin needs to be applied later, but I wasn't sure about how you wanted to do that.

**Comment by aliu139** [comment]
https://docs.google.com/document/d/1DAn4SpmezkgO9fHelpSgcuqkQQFpDvj6gUI2mZrK8VY/edit#

File: public/js/admin/checkin.js [hunk 7]
```diff
@@ -6,6 +6,45 @@ app.controller('checkin.ctrl', ['$scope', '$http', function ($scope, $http) {
...
+        target: document.querySelector("#camera video"),
+        activator: document.querySelector("#camera button"),
+        decoder: qr.decodeFromCamera
+    }];
+
+    //In case we have more than one stream
+    elems.forEach(function(e) {
+        e.activator.onclick = function(r) {
+            //Stop any default behavior associated with buttons
+            r && r.preventDefault(),
```
Comment (mahsu) [inline public/js/admin/checkin.js]: I would avoid using the comma syntax. I think it's unnecessarily confusing with little improvement in readability.

@TrevorEdwards @julianahhong thoughts?

File: public/js/admin/checkin.js [hunk 8]
```diff
@@ -6,6 +6,45 @@ app.controller('checkin.ctrl', ['$scope', '$http', function ($scope, $http) {
...
     $scope.inputSearch = "";
 
+    //For QR scanning
+    //Creates QRCodeDecoder Object
+    var qr = new QCodeDecoder;
+    var video = document.getElementById('camera');
+
+    //Ensures that Canvas is supported in the browser
+    if (!qr.isCanvasSupported() || !qr.hasGetUserMedia()){
+        throw alert("Your browser doesn't match the required specs."),
```
Comment (mahsu) [inline public/js/admin/checkin.js]: IMO

````
alert("Your browser doesn't match the required specs.");
throw new Error("Canvas and getUserMedia are required");
````
Looks cleaner

File: public/js/admin/checkin.js [hunk 9]
```diff
@@ -6,6 +6,45 @@ app.controller('checkin.ctrl', ['$scope', '$http', function ($scope, $http) {
...
+            if (e){
+                throw e;
+            }
+
+            //If decode works, then this will alert.
+            //Change this check-in logic later
+            alert("Just decoded: " + r);
+
+            //Reload to kill the stream
+            location.reload();
```
Comment (mahsu) [inline public/js/admin/checkin.js]: Is this required? What happens if we don't reload?

File: views/admin/checkin.jade [hunk 10]
```diff
@@ -7,9 +7,19 @@ block append vars
     - bodyAttr = {"data-ng-app": "brh.controllers"}
 
 block append admin_content
+    script(src='//cirocosta.github.io/qcode-decoder/build/qcode-decoder.min.js')
```
Comment (mahsu) [inline views/admin/checkin.jade]: append js to the footer.

File: public/js/admin/checkin.js (hunk 9 above)
Comment (aliu139) [inline public/js/admin/checkin.js]: The library is set to only do one capture. After it decodes a single image, the stream remains but is completely inactive.

File: public/js/admin/checkin.js (hunk 7 above)
Comment (aliu139) [inline public/js/admin/checkin.js]: Changed to just be a ;

File: public/js/admin/checkin.js (hunk 8 above)
Comment (aliu139) [inline public/js/admin/checkin.js]: Noted and changed

File: views/admin/checkin.jade (hunk 10 above)
Comment (aliu139) [inline views/admin/checkin.jade]: Done

**Comment by TrevorEdwards** [comment]
sorry this review took so long but I'm testing it out now and it seems like it's only reading qr codes if they are showing and in focus as soon as I start scanning. Otherwise, I have to refresh and try again. I'll try to see if I can figure out why / change it to make this more intuitive

**Comment by TrevorEdwards** [comment]
@aliu139 I changed it to non-once and set the deactivators to refresh the page to fix that issue and make the deactivator work a little better. Feel free to merge if the changes look ok to you.

**Comment by aliu139** [comment]
Looks great to me!

**Pull request merged**